# <h1>Homework 2

# <h1> Problem 1
We have been given the following function <br/>
$ f = 2x_{1}^2 - 4x_{1}x_{2} + 1.5x_{2}^2 + x_{2}
$
<br/>
First, let's find the stationary point <br/>
$i.e \; (x_{1},x_{2})$ where $\frac{\partial f}{\partial x_1} = 0$ and $\frac{\partial f}{\partial x_2} = 0$
simultaneously
<br/>
Thus, $\frac{\partial f}{\partial x_1} = 4x_{1} - 4x_{2} \rightarrow (1)$
<br/> and <br/>
$\frac{\partial f}{\partial x_2} = -4x_{1} + 3x_{2} + 1 \rightarrow (2)$
<br/> Solving $(1)$ and $(2)$ simultaneously we get,
$(x_{1},x_{2}) = [1,1]$
<br/> Thus, stationary point of the function is $[1,1]$
<br/> To find whether a point is saddle point for the function or not, we need to compute its hessian
at the point.<br/>
Therefore, $H = \begin{bmatrix}
   \frac{\partial^2 f}{\partial x_{1}^2} & \frac{\partial^2 f}{\partial x_{1}x_{2}}\\
   \frac{\partial^2 f}{\partial x_{2}x_{1}} & \frac{\partial^2 f}{\partial x_{2}^2}
  \end{bmatrix}$
 <br/>
 $H = \begin{bmatrix}
 4 & -4\\
 -4 & 3
 \end{bmatrix}$

In [5]:
import numpy as np
from numpy import linalg as la
eig,v = la.eig(np.array([[4 ,-4],[-4,3]]))
print(eig)

[ 7.53112887 -0.53112887]


<br/> The resulting Hessian is an indefinite matrix, since the matrix eigen values are both positive and negative.
Hence, for this function all stationary points are saddle points.

<h1> Problem 2<br/>
<h2> Part a

In this problem, we are asked to find the point on plane
$x_{1} + 2x_{2} + 3x_{3} = 1$ closest to the point $(- 1,0,1)^{T}$
<br/> Firstly, to convert this into an unconstrained problem we substitute <br/>
$x_{1} = 1 - 2x_{2} - 3x_{3}$ <br/>
Thus, distance of point from plane can be represented as <br/>
$D = \sqrt{(-1 -1 + 2x_{2} + 3x_{3})^{2} + (0-2x_{2})^{2} + (1-3x_{3})^{2} }$
<br/>Thus, our job is to
<br/> $minimize \; D$

In [7]:
import numpy as np
import math
from scipy.optimize import minimize

In [21]:
def cost_func(x):
    return math.sqrt((-2+2*x[0]+3*x[1])**2+(2*x[0])**2+(1-3*x[1])**2)

In [22]:
x0 = np.array([-1,-1])
sol = minimize(cost_func,x0)

In [29]:
print("[x2,x3] = ",sol.x)
print("x1 = ",1-2*sol.x[0]-3*sol.x[1])

[x2,x3] =  [0.16666675 0.44444444]
x1 =  -0.6666668279503635


Thus the point on plane $x_{1} + 2x_{2} + 3x_{3} = 1$ closest to the point $(- 1,0,1)^{T}$ is
$\begin{bmatrix}
x_{1}\\
x_{2}\\
x_{3}
\end{bmatrix}$ = $
\begin{bmatrix}
-0.6666\\
0.1666\\
0.4444
\end{bmatrix}$